### PR TITLE
beta6 upgrade

### PR DESCRIPTION
**Pull request by reqshark**
@nickdesaulniers, the upgrade is done.

addresses #102 and half of #103

### CODE REVIEW

**Comment by thelinuxlich** [comment]
great!

File: binding.gyp [hunk 1]
```diff
@@ -120,24 +120,36 @@
...
       'target_name': 'node_nanomsg',
+      'dependencies': [ 'nanomsg', ],
       'cflags': [ '-Wall -Werror -Wno-unused' ],
       'cflags_cc': ['-fexceptions'],
       'conditions': [
         ['OS=="mac"', {
-          'ldflags': ['-ldtrace'],
-          'libraries': ['-ldtrace' ],
+          'ldflags': [ '-L<(PRODUCT_DIR) -ldtrace' ],
+          'libraries': [ '-ldtrace', '-L<(PRODUCT_DIR)' ],
```
Comment (nickdesaulniers) [inline binding.gyp]: we don't mean to be linking in dtrace, right?

File: macosx.gypi [hunk 2]
```diff
@@ -3,23 +3,30 @@
...
+            'NN_USE_KQUEUE',
             'NN_USE_IFADDRS',
+            'NN_USE_EVENTFD',
+            'NN_HAVE_MSG_CONTROL'
         ],
         'include_dirs': [
           'deps/nanomsg/src',
         ],
+        'cflags': [
+          '-Wall -Werror -Wno-unused -g -O2 -D_THREAD_SAFE'
```
Comment (nickdesaulniers) [inline macosx.gypi]: Don't need debug symbols.  Can we crank optimizations up to 11...I mean O3? :rocket:

File: binding.gyp (hunk 1 above)
Comment (reqshark) [inline binding.gyp]: nope, i'll take it out

File: macosx.gypi (hunk 2 above)
Comment (reqshark) [inline macosx.gypi]: good idea!

File: macosx.gypi [hunk 3]
```diff
@@ -3,23 +3,30 @@
...
+            'NN_USE_KQUEUE',
             'NN_USE_IFADDRS',
+            'NN_USE_EVENTFD',
+            'NN_HAVE_MSG_CONTROL'
         ],
         'include_dirs': [
           'deps/nanomsg/src',
         ],
+        'cflags': [
+          '-g -O3 -D_THREAD_SAFE'
```
Comment (nickdesaulniers) [inline macosx.gypi]: drop `-g`, we don't need debug symbols :oncoming_police_car:

**Comment by nickdesaulniers** [comment]
I get an error with one test:
```
git submodule sync
git submodule update
npm i
npm t
...
/Users/Nicholas/code/node/nanomsg/node_modules/tape/index.js:75
        throw err
                  ^
Error: stream.push() after EOF
    at readableAddChunk (_stream_readable.js:149:15)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at Socket.<anonymous> (/Users/Nicholas/code/node/nanomsg/lib/index.js:207:24)
```

**Comment by reqshark** [comment]
are you sure you're pulling from the right submodule? what's the last commit hash on that submodule?

**Comment by nickdesaulniers** [comment]
`* 5515f96 - (HEAD, tag: 0.6-beta) Bump ABI version for 0.6-beta. (4 weeks ago) <Garrett D'Amore>`

**Comment by reqshark** [comment]
you're using node 0.12 aren't you

**Comment by reqshark** [comment]
using node 0.12 i'm able to reproduce your result, apparently returning after the survey test passes (and just before it completes).

```bash
✔  bent@osx /Users/bent.cardan/reqshark/node-nanomsg  [beta6|✔ ] 
12:53 $ node test/survey
TAP version 13
# inproc socket survey
ok 1 (unnamed assert)
ok 2 (unnamed assert)
ok 3 (unnamed assert)
ok 4 (unnamed assert)
/Users/bent.cardan/reqshark/node-nanomsg/node_modules/tape/index.js:75
        throw err
              ^
Error: stream.push() after EOF
    at readableAddChunk (_stream_readable.js:149:15)
    at Socket.Readable.push (_stream_readable.js:126:10)
    at Socket.<anonymous> (/Users/bent.cardan/reqshark/node-nanomsg/lib/index.js:207:24)
```

we need to rewrite node.js survey sockets in light of updates to respondent/survey behavior in libnanomsg. note that what we're testing here calls
```js
socket.survey(msg)
``` 
on a surveyor socket, which is different than calling
```js
socket.send(msg)
```

this is [an inconsistent area of our API](https://github.com/nickdesaulniers/node-nanomsg/blob/b60a5130f7acbc03bacba57faf33212f88d0194a/lib/index.js#L311-L328) that has little or nothing to do with the upgrade. 

one problem i see here is that we're [removing a listener](https://github.com/nickdesaulniers/node-nanomsg/blob/b60a5130f7acbc03bacba57faf33212f88d0194a/lib/index.js#L323) that doesn't exist. I think this type specific API stuff was done back when libnanomsg only partially implemented the surveyor/respondent pattern.

since then the pattern's been improved by the clib, certainly in beta6 and possibly earlier in beta5. let's open a separate issue to track this and not let it prevent the upgrade

**Comment by reqshark** [comment]
hmm, wait why does the survey pass on node 0.12 before the upgrade... 
```js
15:35 $ node test/survey.js 
TAP version 13
# inproc socket survey
ok 1 (unnamed assert)
ok 2 (unnamed assert)
ok 3 (unnamed assert)
ok 4 (unnamed assert)

1..4
# tests 4
# pass  4

# ok
```

that means it has something to do with the upgrade and we'll need to revise our survey socket code

File: lib/index.js [hunk 4]
```diff
@@ -184,7 +184,7 @@ Socket.prototype._receive = function () {
             this._stopPollSend();
             this._stopPollReceive();
             this.emit('survey-timeout');
-            return;
+            return null;
```
Comment (reqshark) [inline lib/index.js]: returning `null` here inside the survey's `send` operation ends of the stream properly instead of getting `EOF`

**Comment by reqshark** [comment]
@nickdesaulniers i went through our survey code and made a small change to how survey's recv returns.

now I think the problem had more to do with streams than survey actually, though it's still interesting to me that we only saw this behavior with the upgrade.

anyway, by pushing a value of `undefined` onto a stream we get unexpected behavior. 

by pushing `null` we end the stream, which is what i think we intend to do here: https://github.com/nickdesaulniers/node-nanomsg/blob/beta6/lib/index.js#L187

**Comment by reqshark** [comment]
dude when you get back from the BrazilJS conf Porto Alegre, this PR needs some :heart: ... after this is merged we need to fix iojs3.. nan 2... i was holding off on nan2 since this changes the clib slightly.

btw hope ur having a good time down there.. :palm_tree: this comment doesnt have anything to do with the PR but i was checking out your [your conference slides](http://nickdesaulniers.github.io/BrazilJS2015/slides/index.html) and was wondering if you could pls explain:

> Prefer task level parallelism (postMessage) to data level parallelism (SharedArrayBuffer)

**Comment by nickdesaulniers** [comment]
k let's do :beers: brother!

**Comment by nickdesaulniers** [comment]
@reqshark , can you publish to npm? :fried_shrimp:

**Comment by reqshark** [comment]
done. now npm install has beta6 for everything < iojs-v2.5.0. 

We'll need to get nan2 updates done for v3+.

**Pull request merged**